File: statsig-php/src/StatsigOptions.php
<?php

namespace Statsig;

class StatsigOptions
{
    public $__ref = null; // phpcs:ignore

    public function __construct(
        ?string $specs_url = null,
        ?string $log_event_url = null,
        ?object $specs_adapter = null,
        ?object $event_logging_adapter = null,
        ?string $environment = null,
        ?int $event_logging_flush_interval_ms = null,
        ?int $event_logging_max_queue_size = null,
        ?int $specs_sync_interval_ms = null,
        ?string $output_log_level = null,
        ?bool $disable_country_lookup = null,
        ?bool $wait_for_country_lookup_init = null,
        ?bool $wait_for_user_agent_init = null,
        ?bool $enable_id_lists = null,
        ?bool $disable_network = null,
        ?string $id_lists_url = null,
        ?int $id_lists_sync_interval_ms = null,
        ?bool $disable_all_logging = null,
        ?int $init_timeout_ms = null,
        ?bool $fallback_to_statsig_api = null,
        ?bool $use_third_party_ua_parser = null,
    ) {
        $ffi = StatsigFFI::get();
        $this->__ref = $ffi->statsig_options_create(
            $specs_url,
            $log_event_url,
            is_null($specs_adapter) ? 0 : $specs_adapter->__ref,
            is_null($event_logging_adapter) ? 0 : $event_logging_adapter->__ref,
            $environment,
            $event_logging_flush_interval_ms ?? -1,
            $event_logging_max_queue_size ?? -1,
            $specs_sync_interval_ms ?? -1,
            $output_log_level ?? null,
            toSafeOptBool($disable_country_lookup),
            toSafeOptBool($wait_for_country_lookup_init),
            toSafeOptBool($wait_for_user_agent_init),
            toSafeOptBool($enable_id_lists),
            toSafeOptBool($disable_network),
            $id_lists_url,
            $id_lists_sync_interval_ms ?? -1,
            toSafeOptBool($disable_all_logging),
            null, // global custom fields
            0, // ob client ref - not implemented in PHP
            0, // data store ref - not implemented in PHP
            $init_timeout_ms ?? -1,
            toSafeOptBool($fallback_to_statsig_api),
            toSafeOptBool($use_third_party_ua_parser)
        );
    }

    public function __destruct()
    {
        if (is_null($this->__ref)) {
            return;
        }

        StatsigFFI::get()->statsig_options_release($this->__ref);
        $this->__ref = null;
    }
}

function toSafeOptBool(?bool $value): int
{
    if (is_null($value)) {
        return -1;
    }

    return $value ? 1 : 0;
}

File: statsig-java/src/main/java/com/statsig/StatsigOptions.java
package com.statsig;

public class StatsigOptions {
  private static class CleaningAction implements Runnable {
    private final long ref;

    CleaningAction(long ref) {
      this.ref = ref;
    }

    @Override
    public void run() {
      StatsigJNI.statsigOptionsRelease(ref);
    }
  }

  private volatile long ref;

  private StatsigOptions(Builder builder) {
    /**
     * WARNING: The order of parameters in this method **MUST MATCH EXACTLY** with the corresponding
     * Rust implementation in `statsig_options_jni.rs`. Any mismatch will cause incorrect values to
     * be passed across the JNI boundary.
     */
    this.ref =
        StatsigJNI.statsigOptionsCreate(
            builder.specsUrl,
            builder.logEventUrl,
            builder.idListsUrl,
            builder.idListsSyncIntervalMs,
            builder.specsSyncIntervalMs,
            builder.eventLoggingFlushIntervalMs,
            builder.eventLoggingMaxQueueSize,
            builder.eventLoggingMaxPendingBatchQueueSize,
            builder.initTimeoutMs,
            builder.environment,
            builder.outputLoggerLevel.getValue(),
            builder.serviceName,
            builder.observabilityClient,
            builder.dataStore,
            builder.outputLoggerProvider,
            builder.proxyConfig,
            builder.enableIDLists,
            builder.waitForCountryLookupInit,
            builder.disableAllLogging,
            builder.waitForUserAgentInit,
            builder.disableNetwork,
            builder.disableUserCountryLookup,
            builder.fallbackToStatsigApi,
            builder.useThirdPartyUAParser);

    ResourceCleaner.register(this, new CleaningAction(ref));
  }

  long getRef() {
    return ref;
  }

  public static class Builder {
    private String specsUrl;
    private String logEventUrl;
    private String idListsUrl;
    private long idListsSyncIntervalMs;
    private long specsSyncIntervalMs;
    private long eventLoggingFlushIntervalMs;
    private long eventLoggingMaxQueueSize;
    private long eventLoggingMaxPendingBatchQueueSize;
    private long initTimeoutMs;
    private String environment;
    private ObservabilityClient observabilityClient;
    private DataStore dataStore;
    private OutputLoggerProvider outputLoggerProvider;
    private ProxyConfig proxyConfig;
    private boolean enableIDLists = false;
    private boolean waitForUserAgentInit = false;
    private boolean waitForCountryLookupInit = false;
    private boolean disableUserCountryLookup = false;
    private boolean disableAllLogging = false;
    private boolean disableNetwork = false;
    private boolean fallbackToStatsigApi = false;
    private OutputLogger.LogLevel outputLoggerLevel = OutputLogger.LogLevel.WARN;
    private String serviceName;
    private boolean useThirdPartyUAParser;

    public Builder setOutputLoggerLevel(OutputLogger.LogLevel level) {
      this.outputLoggerLevel = level;
      OutputLogger.logLevel = level;
      return this;
    }

    public Builder setSpecsUrl(String specsUrl) {
      this.specsUrl = specsUrl;
      return this;
    }

    public Builder setProxyConfig(ProxyConfig proxyConfig) {
      this.proxyConfig = proxyConfig;
      return this;
    }

    public Builder setLogEventUrl(String logEventUrl) {
      this.logEventUrl = logEventUrl;
      return this;
    }

    public Builder setSpecsSyncIntervalMs(long specsSyncIntervalMs) {
      this.specsSyncIntervalMs = specsSyncIntervalMs;
      return this;
    }

    public Builder setEventLoggingFlushIntervalMs(long eventLoggingFlushIntervalMs) {
      this.eventLoggingFlushIntervalMs = eventLoggingFlushIntervalMs;
      return this;
    }

    public Builder setEventLoggingMaxQueueSize(long eventLoggingMaxQueueSize) {
      this.eventLoggingMaxQueueSize = eventLoggingMaxQueueSize;
      return this;
    }

    public Builder setEventLoggingMaxPendingBatchQueueSize(
        long eventLoggingMaxPendingBatchQueueSize) {
      this.eventLoggingMaxPendingBatchQueueSize = eventLoggingMaxPendingBatchQueueSize;
      return this;
    }

    public Builder setDataStore(DataStore dataStore) {
      this.dataStore = dataStore;
      return this;
    }

    public Builder setInitTimeoutMs(long initTimeoutMs) {
      this.initTimeoutMs = initTimeoutMs;
      return this;
    }

    public Builder setEnvironment(String environment) {
      this.environment = environment;
      return this;
    }

    public Builder setDisableAllLogging(boolean disableAllLogging) {
      this.disableAllLogging = disableAllLogging;
      return this;
    }

    public Builder setObservabilityClient(ObservabilityClient observabilityClient) {
      this.observabilityClient = observabilityClient;
      return this;
    }

    public Builder setOutputLoggerProvider(OutputLoggerProvider outputLoggerProvider) {
      this.outputLoggerProvider = outputLoggerProvider;
      return this;
    }

    public Builder setIdListsUrl(String idListsUrl) {
      this.idListsUrl = idListsUrl;
      return this;
    }

    public Builder setIdListsSyncIntervalMs(long idListsSyncIntervalMs) {
      this.idListsSyncIntervalMs = idListsSyncIntervalMs;
      return this;
    }

    public Builder setEnableIDLists(boolean enableIDLists) {
      this.enableIDLists = enableIDLists;
      return this;
    }

    public Builder setServiceName(String serviceName) {
      this.serviceName = serviceName;
      return this;
    }

    public Builder setWaitForUserAgentInit(boolean waitForUserAgentInit) {
      this.waitForUserAgentInit = waitForUserAgentInit;
      return this;
    }

    public Builder setWaitForCountryLookupInit(boolean waitForCountryLookupInit) {
      this.waitForCountryLookupInit = waitForCountryLookupInit;
      return this;
    }

    public Builder setFallbackToStatsigApi(boolean fallbackToStatsigApi) {
      this.fallbackToStatsigApi = fallbackToStatsigApi;
      return this;
    }

    public Builder setDisableUserCountryLookup(boolean disableUserCountryLookup) {
      this.disableUserCountryLookup = disableUserCountryLookup;
      return this;
    }

    public Builder setUseThirdPartyUAParser(boolean useThirdPartyUAParser) {
      this.useThirdPartyUAParser = useThirdPartyUAParser;
      return this;
    }

    public Builder setDisableNetwork(boolean disableNetwork) {
      this.disableNetwork = disableNetwork;
      return this;
    }

    public StatsigOptions build() {
      return new StatsigOptions(this);
    }
  }
}
